Feature/retrieve event

ScottBTR opened the pull request:
### Description of Change ###
Added Functionality that allows Calendar events to be pulled back at either a global or calendar specific level for Android.

### Bugs Fixed ###
Related to issue #4 #5

### API Changes ###

Added:

`public static Task<IReadOnlyList<IEvent>> GetEventsAsync(string calendarId = "")`

### Behavioral Changes ### 

Listview for Calendar display altered to be a Picker,

When picking a calendar the associated events are displayed.

### PR Checklist ###
- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

ScottBTR committed:
Calendar Events able to be displayed for all or a specified calendar

ScottBTR committed:
Get Calendars working in sample for both all or specificed calendar on one page.

ScottBTR committed:
Reorder properties so that functionality comes after all properties.

nickrandolph commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,61 @@
+                ItemsSource="{Binding Calendars}"
+                ItemDisplayBinding="{Binding Name}"
+                SelectedItem="{Binding SelectedCalendar, Mode=TwoWay}"/>
+        <ListView Grid.Row="2"
```
We should use CollectionView for this - hmmmm although I guess they may not be targetting XF4.3 yet. Can you confirm whether we can use CollectionView?

mkieres replied:
I asked Scott to do that but they're not using 4.3 yet, therefore we went with ListView

nickrandolph commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,61 @@
+            <ColumnDefinition/>
+            <ColumnDefinition/>
+          </Grid.ColumnDefinitions>
+          <Button Text="Request Read Access"
```
When this page loads, we should query to see if we have read permissions. If we do, we should change the UI to do something like hide the button and/or indicate which permissions the user has granted

ScottBTR replied:
Was having a look into this one. The only issue at the moment is that PermissionStatus is not public so I'm unable to return the enum result from the function:
internal static Task<PermissionStatus> CheckStatusAsync(PermissionType permission) =>
PlatformCheckStatusAsync(permission);

every sample only seems to use:
internal static async Task RequireAsync(PermissionType permission)
        {
            if (await RequestAsync(permission) != PermissionStatus.Granted)
                throw new PermissionException($"{permission} was not granted.");
        }

(as they have no check permission status calls)

Which leaves me with a few options:
A (bool) - make a call similar to the one above but just say to return true if Granted
B (PermissionType) - Allow permission type to be publicly accessible, Microsoft might not agree.
c (local PermissionType) - Replicate the PermissionType enum in my sample allowing me to return it without exposing their private enum.

nickrandolph commented on Samples/Samples/ViewModel/CalendarViewModel.cs:
```diff
@@ -11,28 +13,77 @@ public class CalendarViewModel : BaseViewModel
 
+        public ObservableCollection<IEvent> Events { get; } = new ObservableCollection<IEvent>();
+
+        ICalendar selectedCalendar;
```
Is this the correct ordering of public/private members?

ScottBTR replied:
I'm gonna say no, correcting now

nickrandolph commented on Samples/Samples/ViewModel/CalendarViewModel.cs:
```diff
@@ -11,28 +13,77 @@ public class CalendarViewModel : BaseViewModel
+                if (selectedCalendar != value)
+                {
+                    selectedCalendar = value;
+                    OnPropertyChanged("SelectedCalendar");
```
use nameof(SelectedCalendar)
better still - if the base viewmodel has it, use SetProperty

ScottBTR replied:
Yep, fixed

mkieres commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,61 @@
-                </ViewCell>
-            </DataTemplate>
-        </ResourceDictionary>
+             Title="Calendars">
```
Probably not essential but I would suggest defining `x:DataType` at the page level and in the `DateTemplate`s

nickrandolph commented on Samples/Samples/ViewModel/CalendarViewModel.cs:
```diff
@@ -11,28 +13,77 @@ public class CalendarViewModel : BaseViewModel
+        async void OnChangeRequestCalendarSpecificEvents(string calendarId)
+        {
+            Events.Clear();
+            var events = await Calendar.GetEventsAsync(calendarId);
```
If calendarId==null will this return events across all calendars, or should it be calling OnClickGetEvents for that?

ScottBTR replied:
Removing method OnClickGetEvents as it was used for the other/deprecated implementation of xaml.

As the method handles a null parameter as getting all, I will alter the function to:
async void OnChangeRequestCalendarSpecificEvents(string calendarId = null)

To make it more obvious, I can add a comment as well if this is still not clear? But if I were should I add that above this function, or within the Calendar class, since this is only a sample app so if people were to f12 into the function to understand what it was doing I'm guessing they would be taken to the Calendar.shared.cs implementation, and thus it may be better to have a comment in there.

mkieres commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,61 @@
+          </Grid.ColumnDefinitions>
+          <Button Text="Request Read Access"
+                  Command="{Binding RequestCalendarReadAccess}"
+                  HeightRequest="100"/>
```
I would probably put this `HeightRequest` on these buttons container `Grid` - however we should investigate why it's not resizing automatically

nickrandolph commented on Xamarin.Essentials/Types/DeviceCalendar.shared.cs:
```diff
@@ -22,9 +22,44 @@ public interface ICalendar
-        public int Id { get; set; }
+        string Id { get; set; }
+
+        // Calendar Id this event is for
```
comment is self explanatory, remove

ScottBTR replied:
Yep, removed.

nickrandolph commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -52,5 +52,60 @@ static bool IsCalendarReadOnly(CalendarAccess accessLevel)
+                    Title = cur.GetString(eventsProjection.IndexOf(CalendarContract.Events.InterfaceConsts.Title)),
+                    Description = cur.GetString(eventsProjection.IndexOf(CalendarContract.Events.InterfaceConsts.Description)),
+                    Location = cur.GetString(eventsProjection.IndexOf(CalendarContract.Events.InterfaceConsts.EventLocation)),
+                    AllDay = cur.GetInt(eventsProjection.IndexOf(CalendarContract.Events.InterfaceConsts.AllDay)) == 1,
```
Is the definition for these bool properties that the value == 1 or that it's !=0. Subtle difference but we should adhere to what the docs say

ScottBTR replied:
Due to some documentation I've found online quoting:
"If allDay is set to 1 eventTimezone must be Time#TIMEZONE_UTC and the time must correspond to a midnight boundary."

I may leave this in for now.

ScottBTR committed:
Trial including start and end dates for event requests.

ScottBTR committed:
Allow Date & Time pickers to control event display (also works with specific calendars)

ScottBTR committed:
Revert Linker change.

nickrandolph commented on Samples/Samples/Behaviors/EventToCommandBehaviour.cs:
```diff
@@ -0,0 +1,132 @@
+﻿using System;
```
Please include a comment as to where this code came from

ScottBTR replied:
Added code reference in latest commit see: https://github.com/builttoroam/Essentials/pull/17/commits/ea7116121923e124f1f671fbc1b589b13302158f

nickrandolph replied:
Can you do the same for the BehaviorBase class

ScottBTR replied:
Done

ScottBTR committed:
Provide code reference point

ScottBTR committed:
Added Checkboxes to enable and disable start date/time and end date/time, Also a modal!

Note - Might need help working out why my EventToCommandBehavior isn't working for my checkboxes.

They just use their property {getters/setters}, which I could just use but it was suggested that I use the EventToCommand approach where possible.

You can also get a modal to appear giving you extra information about an event when tapping on an item within the ListView.

nickrandolph commented on Samples/Samples/Converters/CheckBoxEventArgsConverter.cs:
```diff
@@ -0,0 +1,23 @@
+using System.Text;
+using Xamarin.Forms;
+
+namespace Samples.Converters
```
I'm not happy that we have to add these converters - we should be able to construct the UI to avoid these given this is a sample app. Want to avoid including too many additional files etc

ScottBTR replied:
Removed EpochToDateTimeConverter in next PR request. will consider altering StartDatePickersEnabled to be within the setter, the current approach I took was to learn about best practice. But I understand since it's a sample we may want to consider reducing files down the line.

ScottBTR committed:
Added Github repository reference to class pulled from outside source.

nickrandolph commented on Samples/Samples/View/CalendarEventPage.xaml:
```diff
@@ -0,0 +1,68 @@
+                   Text="When:"/>
+            <Label Grid.Column="1"
+                   Grid.Row="2"
+                   Text="{Binding Start, Converter={StaticResource EpochToDateTimeConverter}}"
```
Let's move the conversion into the viewmodel so we can remove the need for these converters

nickrandolph commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,172 @@
+                     TextColor="Black"
+                     HorizontalTextAlignment="Center"/>
+              <Label Grid.Column="1"
+                     Text="{Binding Start, Converter={StaticResource EpochToDateTimeConverter}}"
```
See above comment about converters

nickrandolph commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,172 @@
+        </Grid.ColumnDefinitions>
+        <CheckBox Grid.Row="0"
+                  Grid.Column="0"
+                  IsChecked="{Binding StartDatePickersEnabled}">
```
I'd change this to be 2-way binding and use that to trigger the update. Alternatively, you could introduce a load button and only load calendar items when they click the button.

nickrandolph commented on Xamarin.Essentials/Types/DeviceCalendar.shared.cs:
```diff
@@ -22,11 +22,44 @@ public interface ICalendar
+
+        string Status { get; set; }
+
+        List<Reminder> Reminders { get; set; }
```
For the moment I think we should remove Reminders and RecurranceDate and RecurranceFrequency - I'm not sure whether reminders map well across all platforms and we should be exposing nicer properties for recurring events

ScottBTR replied:
Altered in next PR

mkieres commented on Samples/Samples.Android/Properties/AndroidManifest.xml:
```diff
@@ -1,24 +1,18 @@
+	<uses-feature android:name="android.hardware.location" android:required="false" />
+	<uses-feature android:name="android.hardware.location.gps" android:required="false" />
+	<uses-feature android:name="android.hardware.location.network" android:required="false" />
+	<application android:label="@string/app_name" android:icon="@mipmap/ic_launcher_round" android:roundIcon="@mipmap/ic_launcher_round" android:theme="@style/MainTheme" android:fullBackupContent="@xml/my_backup_rules"></application>
```
Not fussed about this but I would probably try to make sure that this file (or any other for that matter) is nicely formatted so it's easier to read. In this case I would probably apply the same rule as we have for XAML, where each attribute is on a separate line (unless that's not what the essentials team expects)

mkieres commented on Samples/Samples/Behaviors/BehaviorBase.cs:
```diff
@@ -0,0 +1,49 @@
+using System.Text;
+using Xamarin.Forms;
+
+/*
```
Just making sure that we follow their style of comments - I rarely see the usage of `/* */`. Should this be also at placed at the top of the file? I'm thinking about something like this https://github.com/xamarinhq/xamu-infrastructure/blob/598f97d6f5cc68fc603873e913ccc9089dbc60df/src/XamU.Core/Collections/OptimizedObservableCollection.cs

mkieres commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,172 @@
+          <ColumnDefinition/>
+        </Grid.ColumnDefinitions>
+        <Button Text="Request Read Access"
+                Command="{Binding RequestCalendarReadAccess}"
```
Just making sure that we're following the naming convention here - should we add a `Command` suffix to the name of it?

mkieres commented on Samples/Samples/View/CalendarPage.xaml:
```diff
@@ -1,49 +1,172 @@
+        <DatePicker Date="{Binding StartDate}"
+                    MinimumDate="1970-01-01"
+                    IsEnabled="{Binding StartDatePickersEnabled}"
+                    Grid.Row = "0"
```
Formatting: Unnecessary spacing

mkieres commented on Samples/Samples/View/CalendarPage.xaml.cs:
```diff
@@ -1,10 +1,25 @@
+                return;
+
+            var modal = new CalendarEventPage();
+            modal.BindingContext = e.Item as IEvent;
```
I would probably move this cast up to the if statement, as we probably don't want to show an empty page in case the `e.Item` is not an `IEvent`.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -52,5 +53,70 @@ static bool IsCalendarReadOnly(CalendarAccess accessLevel)
+            var calendarSpecificEvent = string.Empty;
+            if (!string.IsNullOrEmpty(calendarId))
+            {
+                calendarSpecificEvent = $"{CalendarContract.Events.InterfaceConsts.CalendarId}={calendarId} AND ";
```
Generally speaking, I'm not a big fan of leaving string literals in code. I would suggest creating a const for `AND` and using it across the board.

I would probably try to create something similar for the `deleted != 1` part, as it's not super obvious what it does. I would probably name this whole part of query something like `NotDeletedEventsWhereClause` - if you can find a better wording go for it.

mkieres reviewed (approved):
See suggestion in comments.
